### PR TITLE
Various RSpec.clear_examples bug fixes.

### LINKED ISSUE #2367: RSpec clear_examples almost works in a complex setup with Runner.run in a loop.
In the [Logstash](https://github.com/elastic/logstash) product we use rake and rspec to run many specs on the main project and the 100 or so plugins that have their own rspec suite.

We thank the rspec community for their continued work on this important library.

When run individually each plugin spec suite runs fine with the set of helpers required correctly.

If I attempt to automate these with a loop using `RSpec::Core::Runner.run(rspec_args)` and `RSpec.clear_examples` it almost works.

1) the deprecation warnings are not cleared and continue to be reported in a cumulative manner (i know but we have tech debt and a smallish team). I was able to use `--deprecation-out` to reduce the output in the console.
2) the shared_examples from `require ..../shared_examples` are removed by the `world.reset` method doing `@shared_example_group_registry = nil` NOTE: commenting this line out worked for me.
3) the `--out` setting did not work because the IO is closed after the first loop iteration and `clear_examples` does not reopen the IO. I had to pipe the stdout output to a file.

It is important in our case to have the output go to a file because with JRuby stack traces etc. we can't always scroll back in the console.

It is debatable whether the randomized seed should be reset by `clear_examples`.

I tried to change some code with an eye to a PR (its always nice to get fixes with issues, we know) but the finding of how to clear the calls in the `DeprecationReporterBuffer` defeated me. 

May I suggest a method called something like `reset_before_rerun` that is `clear_examples` and more? It would better reveal the intention of the functionality.

Previous discussion - https://github.com/rspec/rspec-core/issues/1700

### PR DESCRIPTION
**Pull request by xaviershay**
Fixes #2367 

See individual commit messages.

This doesn't consider whether we should change/alias the name of `clear_examples` as proposed in linked ticket. We probably do want to do that though.

Probably not interesting to anyone, but I tried streaming my process of making this PR. Took me longer than I thought! Got some weird audio feedback issues so may not be too watchable...

* https://www.twitch.tv/xaviershay/v/114691294
* https://www.twitch.tv/xaviershay/v/114813777
* https://www.twitch.tv/xaviershay/v/114825637

### CODE REVIEW

File: lib/rspec/core.rb [hunk 1]
```diff
@@ -69,7 +69,7 @@ def self.reset
   # same process.
   def self.clear_examples
     world.reset
-    configuration.reporter.reset
+    configuration.reset_reporter
```
Comment (myronmarston) [inline lib/rspec/core.rb]: Is `Reporter#reset` no longer used?  If so, can we remove it?

File: lib/rspec/core/configuration.rb [hunk 2]
```diff
@@ -484,6 +484,11 @@ def reset
         @formatter_loader = nil
       end
 
+      def reset_reporter
```
Comment (JonRowe) [inline lib/rspec/core/configuration.rb]: This either needs some docs or a `# @private` to satisfy our documentation coverage check :)

**Comment by JonRowe** [comment]
> It's weird that reporter belongs to configuration. It mixes state with
configuration. This means that we have to do a "partial" configuration
reset when we want to "reset except configuration", which suggests to me
that we'll probably have more issues here in the future.

Yeah maybe we need to rework how we store the state, maybe we need a "Reporter" and a contained "RunState" or similar.

File: lib/rspec/core/configuration.rb (hunk 2 above)
Comment (myronmarston) [inline lib/rspec/core/configuration.rb]: And I think we should be intentional about it: if we add docs, we are declaring it part of our SemVer public API and committing to support it.  I don't see this needing to be part of our SemVer public API so would recommend `# private`.

File: spec/rspec/core/formatters/base_text_formatter_spec.rb [hunk 3]
```diff
@@ -8,17 +8,20 @@
     let(:output_to_close) { File.new("./output_to_close", "w") }
     let(:formatter) { described_class.new(output_to_close) }
 
-    it 'does not close an already closed output stream' do
+    after do
+      # Windows appears to not let the `:isolated_directory` shared group cleanup if
+      # the file isn't closed, so we need to use `and_call_original` here.
```
Comment (JonRowe) [inline spec/rspec/core/formatters/base_text_formatter_spec.rb]: This comment isn't quite right anymore :)

**Comment by xaviershay** [comment]
still have a couple of minor things to iron out (including responding to feedback), but please do take a look at the commit messages. These changes did my head in...

File: lib/rspec/core.rb (hunk 1 above)
Comment (xaviershay) [inline lib/rspec/core.rb]: fixed

**Comment by xaviershay** [comment]
ready for re-review! This addresses all 3 issues from #2367

**Comment by xaviershay** [comment]
remind me: where did we end up on policy for `CHANGELOG` additions in the PR? This would be a bugfix addition I think.

**Comment by JonRowe** [comment]
@xaviershay the policy was more about not pushing people to do the changelog entries themselves, but we've been adding our own changelog entries in PRs

**Comment by JonRowe** [comment]
This just needs a fix for the ffi issue on 1.9.3 on appveyor and it's good to go

**Comment by JonRowe** [comment]
The windows build has been fixed upstream if you want to rebase to get this green, happy to merge as is though.

**Comment by myronmarston** [comment]
I'm going to rebase this in an attempt to make it go green.

**Comment by myronmarston** [comment]
Nevermind, I see I already did that in #2379.  Gonna close this.

**Comment by myronmarston** [comment]
@xaviershay given this got merged in #2379, can you add something to the changelog for it?

**Comment by xaviershay** [comment]
yes I will do this

**Pull request closed without merging**